src/main/java/org/junit/Assert.java

```diff
@@ -955,4 +955,96 @@ public static void assertEquals(Object[] expecteds, Object[] actuals) {
+     * @since 4.13
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends Throwable> T expectThrows(ThrowingRunnable runnable) {
+        return (T) expectThrows(null, Throwable.class, runnable);
+    }
+
+    /**
+     * Asserts that {@code runnable} throws an exception when executed. If it does, the exception
+     * object is returned. If it does not, an {@link AssertionError} is thrown with the given {@code
+     * message}.
+     *
+     * @param message  the identifying message for the {@link AssertionError}
+     * @param runnable A function that is expected to throw an exception when executed
+     * @return The exception thrown by {@code runnable}
+     * @since 4.13
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends Throwable> T expectThrows(String message, ThrowingRunnable runnable) {
+        return (T) expectThrows(message, Throwable.class, runnable);
+    }
+
+    /**
+     * Asserts that {@code runnable} throws an exception of type {@code throwableClass} when
+     * executed. If it does, the exception object is returned. If it does not throw an exception, an
+     * {@link AssertionError} is thrown. If it throws the wrong type of exception, an {@code
+     * AssertionError} is thrown describing the mismatch; the exception that was actually thrown can
+     * be obtained by calling {@link AssertionError#getCause}.
+     *
+     * @param throwableClass the expected type of the exception
+     * @param runnable       A function that is expected to throw an exception when executed
+     * @return The exception thrown by {@code runnable}
+     * @since 4.13
+     */
+    public static <T extends Throwable> T expectThrows(Class<T> throwableClass, ThrowingRunnable runnable) {
+        return expectThrows(null, throwableClass, runnable);
+    }
+
+    /**
+     * Asserts that {@code runnable} throws an exception of type {@code throwableClass} when
+     * executed. If it does, the exception object is returned. If it does not throw an exception, an
+     * {@link AssertionError} is thrown with the given {@code message}. If it throws the wrong type
+     * of exception, an {@code AssertionError} is thrown describing the mismatch; the exception that
+     * was actually thrown can be obtained by calling {@link AssertionError#getCause}.
+     *
+     * @param message        the identifying message for the {@link AssertionError}
+     * @param throwableClass the expected type of the exception
+     * @param runnable       A function that is expected to throw an exception when executed
+     * @return The exception thrown by {@code runnable}
+     * @since 4.13
+     */
+    @SuppressWarnings("unchecked")
+    public static <T extends Throwable> T expectThrows(String message, Class<T> throwableClass, ThrowingRunnable runnable) {
+        try {
+            runnable.run();
+        } catch (Throwable t) {
+            try {
+                return throwableClass.cast(t);
```

Comment:
You don't need JDK8 to operate with functional interfaces like `Supplier<T>` and `Predicate<T>`. Any backport of any subset or superset of those interfaces is basically equivalent. My first draft of this change used a predicate instead of returning a throwable at all. The _only_ major feature that actually requires JDK8 (from the perspective of a library author) is `default` methods on interfaces. However, all of interfaces you've already defined ship with JUnit itself, and cannot be redefined in junit-lambda. So unless you're planning on building something substantial on a lesser-known JDK8 feature like JSR308 type annotations, I'm not sure what the point of junit-lambda is. The JDK8 backwards compatibility story is very, very solid. You're not _supposed_ to need to vend an extension library.